ClientCredentialsKeypairs/AuthenticationService.cs

```diff
@@ -47,38 +48,51 @@ public AuthenticationService(HttpClient client, ClientCredentialsConfiguration c
 
     public async Task SetupToken()
     {
-        var scopes = string.IsNullOrEmpty(Api.Scope) ? Config.Scopes : Api.Scope;
-        var cctr = new ClientCredentialsTokenRequest
+        var scopeString = string.IsNullOrEmpty(Api.Scope) ? Config.Scopes : Api.Scope;
+        var scopes = scopeString.Split(' ');
+        
+        // Remove /connect/token in case the authority config has this appended
+        var authorityUri = Config.Authority.Replace("/connect/token", "");
+        
+        var confidentialClientApp = ConfidentialClientApplicationBuilder.Create(Config.ClientId)
+            .WithOidcAuthority(authorityUri)
+            .WithClientAssertion((AssertionRequestOptions _) => Task.FromResult(BuildClientAssertion(Config.Authority, Config.ClientId)))
+            .WithExperimentalFeatures()
+            .WithHttpClientFactory(new HttpClientContainer(Client))
+            .Build();
+
+        var request = confidentialClientApp.AcquireTokenForClient(scopes);
+        
+        if (Api.UseDpop)
         {
-            Address = Config.Authority,
-            ClientId = Config.ClientId,
-            DPoPProofToken = Api.UseDpop ? BuildDpopAssertion(HttpMethod.Post, Config.Authority) : null,
-            GrantType = OidcConstants.GrantTypes.ClientCredentials,
-            ClientCredentialStyle = ClientCredentialStyle.PostBody,
-            Scope = scopes,
-            ClientAssertion = new ClientAssertion
+            request.WithExtraHttpHeaders(new Dictionary<string, string>
             {
-                Type = OidcConstants.ClientAssertionTypes.JwtBearer,
-                Value = BuildClientAssertion(Config.Authority, Config.ClientId)
-            }
-        };
+                { DPoPHeaderNames.DPoP, BuildDpopAssertion(HttpMethod.Post, Config.Authority) }
+            }).WithProofOfPosessionKeyId(Guid.NewGuid().ToString(), "DPoP");
+        }
 
-        var response = await Client.RequestClientCredentialsTokenAsync(cctr);
-        if (response.IsError)
+        try
         {
-            if (Api.UseDpop && response.Error == OidcConstants.TokenErrors.UseDPoPNonce)
-            {
-                cctr.DPoPProofToken = BuildDpopAssertion(HttpMethod.Post, Config.Authority, nonce: response.DPoPNonce ?? Guid.NewGuid().ToString());
-                response = await Client.RequestClientCredentialsTokenAsync(cctr);
-            }
+            var result = await request
+                .ExecuteAsync();
 
-            if (response.IsError)
-            {
-                throw new Exception($"Unable to get access token: {response.Error}");
-            }
+            _accessToken = result.AccessToken;
         }
+        catch (MsalServiceException e) when (e.ErrorCode == DPoPErrorCode.UseDPoPNonce)
+        {
+            var nonce = e.Headers.GetValues(DPoPHeaderNames.Nonce).FirstOrDefault() ??
+                        throw new Exception("There must be exactly one value for the DPoP-Nonce header");
```

Comment:
```suggestion
                        throw new DPoPException("There must be exactly one value for the DPoP-Nonce header");
```

Lag egen exception, don't throw base class Exception